### ISSUE TITLE
Duplicati stuck on verifying backend data

### ISSUE THREAD

**Issue by Rubydesic** [duplicati/duplicati#3205]
- [x] I have searched open and closed issues for duplicates.

----------------------------------------

## Environment info

- **Duplicati version**: 2.0.3.3_beta_2018-04-02
- **Operating system**: Windows 10 Pro Version 1709 Build 16299.371 
- **Backend**: Google Drive

## Description
Duplicati is stuck on "Verifying Backend Data". This occurs even after I enable the "no-backend-verification" option. This only started happening after the update from the previous beta version.

## Steps to reproduce
1. Start duplicati
2. Start a backup

- **Actual result**:

Duplicati never begins the backup and remains on the "Verifying Backend Data" stage

- **Expected result**:

Duplicati starts a backup.

## Debug log

```May 4, 2018 6:55 PM: Delete operation failed for duplicati-i5a41f5ee55f14c6ea2b3ad7ff2c17191.dindex.zip.aes with FileNotFound, listing contents
May 4, 2018 6:55 PM: Backend event: Delete - Started: duplicati-i5a41f5ee55f14c6ea2b3ad7ff2c17191.dindex.zip.aes ()
May 4, 2018 6:55 PM: Starting - RemoteOperationDelete
May 4, 2018 6:55 PM: removing remote file listed as Temporary: duplicati-i5a41f5ee55f14c6ea2b3ad7ff2c17191.dindex.zip.aes
May 4, 2018 6:55 PM: RemoteOperationDelete took 00:00:47.292
May 4, 2018 6:55 PM: Listing indicates file duplicati-i6f313f495c1c4970b5aed4929bfbb415.dindex.zip.aes is deleted correctly
May 4, 2018 6:54 PM: Delete operation failed for duplicati-i6f313f495c1c4970b5aed4929bfbb415.dindex.zip.aes with FileNotFound, listing contents
May 4, 2018 6:54 PM: Backend event: Delete - Started: duplicati-i6f313f495c1c4970b5aed4929bfbb415.dindex.zip.aes ()
May 4, 2018 6:54 PM: Starting - RemoteOperationDelete
May 4, 2018 6:54 PM: removing remote file listed as Deleting: duplicati-i6f313f495c1c4970b5aed4929bfbb415.dindex.zip.aes
May 4, 2018 6:54 PM: RemoteOperationDelete took 00:00:48.182
May 4, 2018 6:54 PM: Listing indicates file duplicati-i5cc05a8e8b91473dbdf6c53bb152cc9e.dindex.zip.aes is deleted correctly
May 4, 2018 6:54 PM: Delete operation failed for duplicati-i5cc05a8e8b91473dbdf6c53bb152cc9e.dindex.zip.aes with FileNotFound, listing contents
May 4, 2018 6:54 PM: Backend event: Delete - Started: duplicati-i5cc05a8e8b91473dbdf6c53bb152cc9e.dindex.zip.aes ()
May 4, 2018 6:54 PM: Starting - RemoteOperationDelete
May 4, 2018 6:54 PM: removing remote file listed as Deleting: duplicati-i5cc05a8e8b91473dbdf6c53bb152cc9e.dindex.zip.aes
May 4, 2018 6:54 PM: RemoteOperationDelete took 00:01:06.145
May 4, 2018 6:54 PM: Listing indicates file duplicati-i08dfbc2aa8de4208b9d35ddbb5f7f502.dindex.zip.aes is deleted correctly
May 4, 2018 6:53 PM: Delete operation failed for duplicati-i08dfbc2aa8de4208b9d35ddbb5f7f502.dindex.zip.aes with FileNotFound, listing contents
May 4, 2018 6:52 PM: Backend event: Delete - Started: duplicati-i08dfbc2aa8de4208b9d35ddbb5f7f502.dindex.zip.aes ()
May 4, 2018 6:52 PM: Starting - RemoteOperationDelete
May 4, 2018 6:52 PM: removing remote file listed as Deleting: duplicati-i08dfbc2aa8de4208b9d35ddbb5f7f502.dindex.zip.aes
May 4, 2018 6:52 PM: Starting - PreBackupVerify
May 4, 2018 6:52 PM: ExecuteReader: SELECT DISTINCT "Fileset"."ID", "Fileset"."Timestamp" FROM "Fileset", "RemoteVolume" WHERE "RemoteVolume"."ID" = "Fileset"."VolumeID" AND "Fileset"."ID" IN (SELECT "FilesetID" FROM "FilesetEntry") AND ("RemoteVolume"."State" = "Uploading" OR "RemoteVolume"."State" = "Temporary") took 00:00:00.260
May 4, 2018 6:52 PM: Starting - ExecuteReader: SELECT DISTINCT "Fileset"."ID", "Fileset"."Timestamp" FROM "Fileset", "RemoteVolume" WHERE "RemoteVolume"."ID" = "Fileset"."VolumeID" AND "Fileset"."ID" IN (SELECT "FilesetID" FROM "FilesetEntry") AND ("RemoteVolume"."State" = "Uploading" OR "RemoteVolume"."State" = "Temporary")
May 4, 2018 6:52 PM: ExecuteScalarInt64: SELECT SUM("Blockset"."Length") FROM "File", "FilesetEntry", "Blockset" WHERE "File"."ID" = "FilesetEntry"."FileID" AND "File"."BlocksetID" = "Blockset"."ID" AND "FilesetEntry"."FilesetID" = ? AND "File"."BlocksetID" NOT IN (?, ?) took 00:00:00.308
May 4, 2018 6:52 PM: Starting - ExecuteScalarInt64: SELECT SUM("Blockset"."Length") FROM "File", "FilesetEntry", "Blockset" WHERE "File"."ID" = "FilesetEntry"."FileID" AND "File"."BlocksetID" = "Blockset"."ID" AND "FilesetEntry"."FilesetID" = ? AND "File"."BlocksetID" NOT IN (?, ?)
May 4, 2018 6:52 PM: ExecuteScalarInt64: SELECT COUNT(*) FROM "File" INNER JOIN "FilesetEntry" ON "File"."ID" = "FilesetEntry"."FileID" WHERE "FilesetEntry"."FilesetID" = ? AND "File"."BlocksetID" NOT IN (?, ?) took 00:00:00.333
May 4, 2018 6:52 PM: Starting - ExecuteScalarInt64: SELECT COUNT(*) FROM "File" INNER JOIN "FilesetEntry" ON "File"."ID" = "FilesetEntry"."FileID" WHERE "FilesetEntry"."FilesetID" = ? AND "File"."BlocksetID" NOT IN (?, ?)
May 4, 2018 6:52 PM: ExecuteScalarInt64: SELECT "ID" FROM "Fileset" ORDER BY "Timestamp" DESC LIMIT 1 took 00:00:00.000
May 4, 2018 6:52 PM: Starting - ExecuteScalarInt64: SELECT "ID" FROM "Fileset" ORDER BY "Timestamp" DESC LIMIT 1
May 4, 2018 6:52 PM: ExecuteScalarInt64: SELECT COUNT(*) FROM "File" WHERE "BlocksetID" != ? AND "BlocksetID" != ? AND NOT "BlocksetID" IN (SELECT "BlocksetID" FROM "BlocksetEntry") took 00:00:01.898
May 4, 2018 6:52 PM: Starting - ExecuteScalarInt64: SELECT COUNT(*) FROM "File" WHERE "BlocksetID" != ? AND "BlocksetID" != ? AND NOT "BlocksetID" IN (SELECT "BlocksetID" FROM "BlocksetEntry")
May 4, 2018 6:52 PM: ExecuteScalarInt64: SELECT COUNT(*) FROM (SELECT * FROM (SELECT "N"."BlocksetID", (("N"."BlockCount" + 3200 - 1) / 3200) AS "BlocklistHashCountExpected", CASE WHEN "G"."BlocklistHashCount" IS NULL THEN 0 ELSE "G"."BlocklistHashCount" END AS "BlocklistHashCountActual" FROM (SELECT "BlocksetID", COUNT(*) AS "BlockCount" FROM "BlocksetEntry" GROUP BY "BlocksetID") "N" LEFT OUTER JOIN (SELECT "BlocksetID", COUNT(*) AS "BlocklistHashCount" FROM "BlocklistHash" GROUP BY "BlocksetID") "G" ON "N"."BlocksetID" = "G"."BlocksetID" WHERE "N"."BlockCount" > 1) WHERE "BlocklistHashCountExpected" != "BlocklistHashCountActual") took 00:00:02.189
May 4, 2018 6:52 PM: Starting - ExecuteScalarInt64: SELECT COUNT(*) FROM (SELECT * FROM (SELECT "N"."BlocksetID", (("N"."BlockCount" + 3200 - 1) / 3200) AS "BlocklistHashCountExpected", CASE WHEN "G"."BlocklistHashCount" IS NULL THEN 0 ELSE "G"."BlocklistHashCount" END AS "BlocklistHashCountActual" FROM (SELECT "BlocksetID", COUNT(*) AS "BlockCount" FROM "BlocksetEntry" GROUP BY "BlocksetID") "N" LEFT OUTER JOIN (SELECT "BlocksetID", COUNT(*) AS "BlocklistHashCount" FROM "BlocklistHash" GROUP BY "BlocksetID") "G" ON "N"."BlocksetID" = "G"."BlocksetID" WHERE "N"."BlockCount" > 1) WHERE "BlocklistHashCountExpected" != "BlocklistHashCountActual")
May 4, 2018 6:52 PM: ExecuteScalarInt64: SELECT Count(*) FROM (SELECT DISTINCT "BlocksetID", "Index" FROM "BlocklistHash") took 00:00:00.009
May 4, 2018 6:52 PM: Starting - ExecuteScalarInt64: SELECT Count(*) FROM (SELECT DISTINCT "BlocksetID", "Index" FROM "BlocklistHash")
May 4, 2018 6:52 PM: ExecuteScalarInt64: SELECT Count(*) FROM "BlocklistHash" took 00:00:00.052
May 4, 2018 6:52 PM: Starting - ExecuteScalarInt64: SELECT Count(*) FROM "BlocklistHash"
May 4, 2018 6:52 PM: ExecuteReader: SELECT "CalcLen", "Length", "A"."BlocksetID", "File"."Path" FROM (
 SELECT 
 "A"."ID" AS "BlocksetID", 
 IFNULL("B"."CalcLen", 0) AS "CalcLen", 
 "A"."Length"
 FROM
 "Blockset" A
 LEFT OUTER JOIN
 (
 SELECT 
 "BlocksetEntry"."BlocksetID",
 SUM("Block"."Size") AS "CalcLen"
 FROM
 "BlocksetEntry"
 LEFT OUTER JOIN
 "Block"
 ON
 "Block"."ID" = "BlocksetEntry"."BlockID"
 GROUP BY "BlocksetEntry"."BlocksetID"
 ) B
 ON
 "A"."ID" = "B"."BlocksetID"
 
 ) A, "File" WHERE "A"."BlocksetID" = "File"."BlocksetID" AND "A"."CalcLen" != "A"."Length" took 00:00:42.644
May 4, 2018 6:50 PM: Server has started and is listening on 127.0.0.1, port 8200```

**Comment by JonMikelV** [comment]
How are you determining it's stuck - just the progress bar?  If so, that may be a UX issue as described here - though I freely admit this may NOT be what you're experiencing:
https://forum.duplicati.com/t/think-youre-stuck-on-verifying-backend-data/3473

**Comment by BlackEdder0815** [comment]
No, not just the progress bar. The job is hanging now since the 4th of May and there is no additional log entry...
Don't think, that this is related to the progress bar issue. The system is doing simply nothing at the moment. Also no additional back-up jobs are processed at the moment.

**Comment by BlackEdder0815** [comment]
Here is my log entry (I restarted server with enabled debug-log):

```
2018-05-09 11:48:22Z - Profiling: Starting - Running Backup
2018-05-09 11:48:22Z - Profiling: Starting - ExecuteScalarInt64: INSERT INTO "Operation" ("Description", "Timestamp") VALUES (?, ?); SELECT last_insert_rowid();
2018-05-09 11:48:22Z - Profiling: ExecuteScalarInt64: INSERT INTO "Operation" ("Description", "Timestamp") VALUES (?, ?); SELECT last_insert_rowid(); took 00:00:00.006
2018-05-09 11:48:22Z - Profiling: Starting - ExecuteReader: SELECT "Key", "Value" FROM "Configuration"
2018-05-09 11:48:22Z - Profiling: ExecuteReader: SELECT "Key", "Value" FROM "Configuration"  took 00:00:00.000
2018-05-09 11:48:22Z - Profiling: Starting - ExecuteReader: SELECT "Key", "Value" FROM "Configuration"
2018-05-09 11:48:22Z - Profiling: ExecuteReader: SELECT "Key", "Value" FROM "Configuration"  took 00:00:00.000
2018-05-09 11:48:22Z - Profiling: Starting - ExecuteScalarInt64: SELECT COUNT(*) FROM "Block" WHERE "Size" > ?
2018-05-09 11:48:24Z - Profiling: ExecuteScalarInt64: SELECT COUNT(*) FROM "Block" WHERE "Size" > ? took 00:00:01.499
2018-05-09 11:48:24Z - Profiling: Starting - ExecuteReader: SELECT "Key", "Value" FROM "Configuration"
2018-05-09 11:48:24Z - Profiling: ExecuteReader: SELECT "Key", "Value" FROM "Configuration"  took 00:00:00.000
2018-05-09 11:48:24Z - Profiling: Starting - ExecuteReader: SELECT "Key", "Value" FROM "Configuration"
2018-05-09 11:48:24Z - Profiling: ExecuteReader: SELECT "Key", "Value" FROM "Configuration"  took 00:00:00.000
2018-05-09 11:48:24Z - Profiling: Starting - ExecuteReader: SELECT "CalcLen", "Length", "A"."BlocksetID", "File"."Path" FROM (
SELECT
    "A"."ID" AS "BlocksetID",
    IFNULL("B"."CalcLen", 0) AS "CalcLen",
    "A"."Length"
FROM
    "Blockset" A
LEFT OUTER JOIN
    (
        SELECT
            "BlocksetEntry"."BlocksetID",
            SUM("Block"."Size") AS "CalcLen"
        FROM
            "BlocksetEntry"
        LEFT OUTER JOIN
            "Block"
        ON
            "Block"."ID" = "BlocksetEntry"."BlockID"
        GROUP BY "BlocksetEntry"."BlocksetID"
    ) B
ON
    "A"."ID" = "B"."BlocksetID"

) A, "File" WHERE "A"."BlocksetID" = "File"."BlocksetID" AND "A"."CalcLen" != "A"."Length"
2018-05-09 11:48:25Z - Profiling: ExecuteReader: SELECT "CalcLen", "Length", "A"."BlocksetID", "File"."Path" FROM (
SELECT
    "A"."ID" AS "BlocksetID",
    IFNULL("B"."CalcLen", 0) AS "CalcLen",
    "A"."Length"
FROM
    "Blockset" A
LEFT OUTER JOIN
    (
        SELECT
            "BlocksetEntry"."BlocksetID",
            SUM("Block"."Size") AS "CalcLen"
        FROM
            "BlocksetEntry"
        LEFT OUTER JOIN
            "Block"
        ON
            "Block"."ID" = "BlocksetEntry"."BlockID"
        GROUP BY "BlocksetEntry"."BlocksetID"
    ) B
ON
    "A"."ID" = "B"."BlocksetID"

) A, "File" WHERE "A"."BlocksetID" = "File"."BlocksetID" AND "A"."CalcLen" != "A"."Length"  took 00:00:00.467
2018-05-09 11:48:25Z - Profiling: Starting - ExecuteScalarInt64: SELECT Count(*) FROM "BlocklistHash"
2018-05-09 11:48:25Z - Profiling: ExecuteScalarInt64: SELECT Count(*) FROM "BlocklistHash" took 00:00:00.000
2018-05-09 11:48:25Z - Profiling: Starting - ExecuteScalarInt64: SELECT Count(*) FROM (SELECT DISTINCT "BlocksetID", "Index" FROM "BlocklistHash")
2018-05-09 11:48:25Z - Profiling: ExecuteScalarInt64: SELECT Count(*) FROM (SELECT DISTINCT "BlocksetID", "Index" FROM "BlocklistHash") took 00:00:00.000
2018-05-09 11:48:25Z - Profiling: Starting - ExecuteScalarInt64: SELECT COUNT(*) FROM (SELECT * FROM (SELECT "N"."BlocksetID", (("N"."BlockCount" + 3200 - 1) / 3200) AS "BlocklistHashCountExpected", CASE WHEN "G"."BlocklistHashCount" IS NULL THEN 0 ELSE "G"."BlocklistHashCount" END AS "BlocklistHashCountActual" FROM (SELECT "BlocksetID", COUNT(*) AS "BlockCount" FROM "BlocksetEntry" GROUP BY "BlocksetID") "N" LEFT OUTER JOIN (SELECT "BlocksetID", COUNT(*) AS "BlocklistHashCount" FROM "BlocklistHash" GROUP BY "BlocksetID") "G" ON "N"."BlocksetID" = "G"."BlocksetID" WHERE "N"."BlockCount" > 1) WHERE "BlocklistHashCountExpected" != "BlocklistHashCountActual")
2018-05-09 11:48:25Z - Profiling: ExecuteScalarInt64: SELECT COUNT(*) FROM (SELECT * FROM (SELECT "N"."BlocksetID", (("N"."BlockCount" + 3200 - 1) / 3200) AS "BlocklistHashCountExpected", CASE WHEN "G"."BlocklistHashCount" IS NULL THEN 0 ELSE "G"."BlocklistHashCount" END AS "BlocklistHashCountActual" FROM (SELECT "BlocksetID", COUNT(*) AS "BlockCount" FROM "BlocksetEntry" GROUP BY "BlocksetID") "N" LEFT OUTER JOIN (SELECT "BlocksetID", COUNT(*) AS "BlocklistHashCount" FROM "BlocklistHash" GROUP BY "BlocksetID") "G" ON "N"."BlocksetID" = "G"."BlocksetID" WHERE "N"."BlockCount" > 1) WHERE "BlocklistHashCountExpected" != "BlocklistHashCountActual") took 00:00:00.132
2018-05-09 11:48:25Z - Profiling: Starting - ExecuteScalarInt64: SELECT COUNT(*) FROM "File" WHERE "BlocksetID" != ? AND "BlocksetID" != ? AND NOT "BlocksetID" IN (SELECT "BlocksetID" FROM "BlocksetEntry")
2018-05-09 11:48:25Z - Profiling: ExecuteScalarInt64: SELECT COUNT(*) FROM "File" WHERE "BlocksetID" != ? AND "BlocksetID" != ? AND NOT "BlocksetID" IN (SELECT "BlocksetID" FROM "BlocksetEntry") took 00:00:00.075
2018-05-09 11:48:25Z - Profiling: Starting - ExecuteReader: SELECT DISTINCT "Fileset"."ID", "Fileset"."Timestamp" FROM "Fileset", "RemoteVolume" WHERE "RemoteVolume"."ID" = "Fileset"."VolumeID" AND "Fileset"."ID" IN (SELECT "FilesetID" FROM "FilesetEntry")  AND ("RemoteVolume"."State" = "Uploading" OR "RemoteVolume"."State" = "Temporary")
2018-05-09 11:48:25Z - Profiling: ExecuteReader: SELECT DISTINCT "Fileset"."ID", "Fileset"."Timestamp" FROM "Fileset", "RemoteVolume" WHERE "RemoteVolume"."ID" = "Fileset"."VolumeID" AND "Fileset"."ID" IN (SELECT "FilesetID" FROM "FilesetEntry")  AND ("RemoteVolume"."State" = "Uploading" OR "RemoteVolume"."State" = "Temporary") took 00:00:00.343
2018-05-09 11:48:25Z - Profiling: Starting - PreBackupVerify
2018-05-09 11:48:25Z - Profiling: Starting - RemoteOperationList
2018-05-09 11:48:25Z - Information: Backend event: List - Started:  ()
2018-05-09 11:48:29Z - Information: Backend event: List - Completed:  (98 Bytes)
2018-05-09 11:48:29Z - Profiling: RemoteOperationList took 00:00:03.490
2018-05-09 11:48:29Z - Profiling: Starting - ExecuteReader: SELECT DISTINCT "Name", "State" FROM "Remotevolume" WHERE "Name" IN (SELECT "Name" FROM "Remotevolume" WHERE "State" IN ("Deleted", "Deleting")) AND NOT "State" IN ("Deleted", "Deleting")
2018-05-09 11:48:29Z - Profiling: ExecuteReader: SELECT DISTINCT "Name", "State" FROM "Remotevolume" WHERE "Name" IN (SELECT "Name" FROM "Remotevolume" WHERE "State" IN ("Deleted", "Deleting")) AND NOT "State" IN ("Deleted", "Deleting") took 00:00:00.003
2018-05-09 11:48:29Z - Information: removing remote file listed as Deleting: duplicati-20180406T230001Z.dlist.zip.aes
2018-05-09 11:48:29Z - Profiling: Starting - RemoteOperationDelete
2018-05-09 11:48:29Z - Information: Backend event: Delete - Started: duplicati-20180406T230001Z.dlist.zip.aes (1,42 MB)
```

And form this point on, duplicati is doing ... nothing.
CPU is at 0%.

**Comment by piegamesde** [comment]
I have a similar issue: "Verifying backend data ..." takes forever. The "Current file" changes at most a few times per minute, even for small files, which would take me weeks to finish checking at that rate. Also, the file progress bar above the "current file" only shows 0% or 100% and the progress bar on top doesn't display any progress (I think it may be at 0%).

**Comment by hamlet42** [comment]
I had an exactly similar case after I had upgraded to v2.0.3.6-2.0.3.6_canary_2018-04-23 and attributed it to this canary for this reason. Used all the available repair options to no avail. So finally, I deleted the backup and downgraded to the prior canary build.

**Comment by KommX** [comment]
I think I ran into the same problem with 2.0.4.16.
Can I contribute any logs or something else to debug this issue?

**Comment by mapl** [comment]
Have the same issue with "Verifying backend data ..." stuck forever with 2.0.4.15_canary_2019-02-06 and .16

**Comment by KommX** [comment]
Even a downgrade to 2.0.4.14 didn't help in my case.

The last lines of the live verbose logs say:
```
Apr 11, 2019 8:10 AM: Re-creating missing index file for duplicati-bddefbd2ed7af45b1ba7d905486575691.dblock.zip.aes
Apr 11, 2019 8:10 AM: Backend event: List - Completed: (9.93 KB)
Apr 11, 2019 8:10 AM: Backend event: List - Started: ()
Apr 11, 2019 8:10 AM: Including path as no filters matched: \\PATH
```
Now one of my CPU cores is at 100 % with no disk activity and no network activity.

@kenkendk @warwickmm 
Does it make sense to comment in this old issue or should I rather open a new issue?

**Comment by Roadrunneruk** [comment]
I have the same thing, had a 3 TB backup which was made about 1 year ago,
I  updated it once a month but out of the blue it would go and say index
files are missing and after 7 days of replace missing index files it
decided to delete the backup and start ago. This time backup up 9GB of
files, it completed with no erro,r but the next time i staredt the backup
it says re creating missing index files. Cpu sits at about 15%.

   - 11 Apr 2019 10:32: CommitAfterUpload took 0:00:00:00.231
   - 11 Apr 2019 10:32: Starting - CommitAfterUpload
   - 11 Apr 2019 10:32: Backend event: Put - Completed:
   duplicati-i7b6a08edc8764ffea9d0b1f69e06d321.dindex.zip.aes (512.01 KB)
   - 11 Apr 2019 10:32: Uploaded 512.01 KB in 00:00:00.0221409, 22.58 MB/s
   - 11 Apr 2019 10:32: Backend event: Put - Started:
   duplicati-i7b6a08edc8764ffea9d0b1f69e06d321.dindex.zip.aes (512.01 KB)
   - 11 Apr 2019 10:32: ExecuteReader: SELECT DISTINCT "Hash", "Size" FROM
   "Block" WHERE "VolumeID" = 20 took 0:00:00:00.000
   - 11 Apr 2019 10:32: Starting - ExecuteReader: SELECT DISTINCT "Hash",
   "Size" FROM "Block" WHERE "VolumeID" = 20
   - 11 Apr 2019 10:32: ExecuteScalarInt64: INSERT INTO "Remotevolume"
   ("OperationID", "Name", "Type", "State", "Size", "VerificationCount",
   "DeleteGraceTime") VALUES (3,
   "duplicati-i057196d383f04c4b83941a0a9f04945a.dindex.zip.aes", "Index",
   "Temporary", -1, 0, 0); SELECT last_insert_rowid(); took 0:00:00:00.000
   - 11 Apr 2019 10:32: Starting - ExecuteScalarInt64: INSERT INTO
   "Remotevolume" ("OperationID", "Name", "Type", "State", "Size",
   "VerificationCount", "DeleteGraceTime") VALUES (3,
   "duplicati-i057196d383f04c4b83941a0a9f04945a.dindex.zip.aes", "Index",
   "Temporary", -1, 0, 0); SELECT last_insert_rowid();
   - 11 Apr 2019 10:32: Re-creating missing index file for
   duplicati-bf006e717fbf447f2a9eb44eeedb88a9a.dblock.zip.aes
   - 11 Apr 2019 10:32: CommitAfterUpload took 0:00:00:00.150
   - 11 Apr 2019 10:32: Starting - CommitAfterUpload
   - 11 Apr 2019 10:32: Backend event: Put - Completed:
   duplicati-i3cc9c692f6924e02bd547bfe14e4184f.dindex.zip.aes (533.15 KB)

LuCI - Lua Configuration Interface </cgi-bin/luci>
<#DAB4FAD8-2DD7-40BB-A1B8-4E2AA1F9FDF2>

**Comment by Roadrunneruk** [comment]
LuCI - Lua Configuration Interface <http:///cgi-bin/luci>
<#m_-5487029535943889985_m_-4018210248387013333_DAB4FAD8-2DD7-40BB-A1B8-4E2AA1F9FDF2>

**Comment by Shellfishgene** [comment]
I have the same problem with a backup to OneDrive. It normally takes about 3-4 minutes to finish, and runs every day. About every 4th or 5th day, it get's stuck on "verifying backend data". I can't cancel it either, I have to kill Duplicati in the task manager. Upon running again, it mostly finishes normally in a few minutes.

**Comment by woutertinusf19** [comment]
I also have this issue with the latest release (2.0.81). I dumped the server process and debugged it using Visual Studio, which shows this state:

![image](https://github.com/user-attachments/assets/a1473cb4-f08d-4319-850f-ad31b4e8d666)

I can provide the full crash dump on request (300MB+). But I guess it's either just waiting infinitly for a file to become available that's not going to become available anymore (for whatever reason, so it should just fail the verification), or it's some kind of deadlock.

**Comment by kenkendk** [comment]
@woutertinusf19 Thanks a bunch for that picture!

From what I can read, the problem is actually in `DirectStreamLink`. It seems to be in the read phase but there does not appear to be any threads providing data.

Without further information, I would guess that the writer has somehow crashed but not closed their end of the stream. Since the reader does not terminate, we never see the underlying error.

Looking at the purpose of `DirectStreamLink` we can just delete it and use `CopyAsync()` to propagate between the streams, as that will make it multithreaded when needed (and use thread overhead when waiting for IO).

**Comment by kenkendk** [comment]
@woutertinusf19 If you get the error regularly, can you perhaps try to set `--disable-piped-streaming=true` ?
This should fully disable the use of `DirectStreamLink`, so at least the hang you have documented is no longer possible.
Perhaps it "just works" or we get a peek at the underlying issue.

**Comment by kenkendk** [comment]
I was looking through the code for the piped setup, and it only offers a minor performance boost if the data is read super fast (like a local file). For a network stream, the CPU is more likely idle during the process.

I think the setup can be built with the `async` pattern where the reader stream will "split" the data so the hashing is done in parallel with the decryption. Using the `async` pattern will also reduce the risk that we end up in the stuck state.

The idea is to have a `HashCalculatingStream` that passes anything read/write through, but also forwards the data to a hasher process, perhaps using channels. This way the core path is from data -> decrypt -> file and the hashing does not add latency. After completing the stream, the hash can be collected.

**Comment by woutertinusf19** [comment]
I have set `disable-piped-streaming` to `true` now on the problematic machine. I'll keep you posted with my findings.

**Comment by kenkendk** [comment]
I had a hunch that the optimization may yield results, but the overhead in thread switching may also slow it down.
I tested it on as small ~5GiB backup destination with the following results:
```
> time /Duplicati.CommandLine test  all --no-encryption   
6.41s user 2.01s system 164% cpu 5.111 total

> time ./Duplicati.CommandLine test  all --no-encryption --disable-piped-streaming  
5.14s user 1.91s system 115% cpu 6.116 total
```

So on my test system, disabling the optimization actually increases performance (in the verification function) by 20%, while *also* reducing CPU usage.

**Comment by woutertinusf19** [comment]
I'm happy to report that with `disable-piped-streaming` the backup was a lot more stable. Where it used to hang every 3-4 days, it now lasted 3 weeks. But over the weekend it got stuck again:

Threads view:
![image](https://github.com/user-attachments/assets/68ed3a64-ed9e-4c64-80dd-9507023f3b05)
Tasks view:
![image](https://github.com/user-attachments/assets/47f519cc-6ba4-44b3-9c43-3018a690953c)

Now we're actually seeing some FTP backend stuff there. It's stuck in a StreamReadHelper that doesn't seem to be making any progress. When I add native code it shows it hanging on Socket.Receive

![image](https://github.com/user-attachments/assets/e522c98e-93e1-409d-ac9b-2c1c27424da0)

I guess a timeout would come in handy?

**Comment by woutertinusf19** [comment]
I noticed that there is some current work happening on in regards to replacing the legacy FTP backend (i.e. https://github.com/duplicati/duplicati/pull/5589/files), which is great, but some protection against misbehaving backends in general might also be worthwhile.

**Comment by marceloduplicati** [comment]
@woutertinusf19 

The PR is essentially removing the legacy FTP code so in future releases all FTP backups will be processed with the FluentFtp lib, it still has some legacy option parsing compatibility issues and its WIP.

Its worth noting that even at your version 2.0.8.1 you can configure the backup using the **FTP (Alternative)** backend that already uses FluentFtp rather than the deprecated FtpWebRequest so that might be a valuable test to assert if the issue is resolved altogether with the new lib.

**Comment by kenkendk** [comment]
> I guess a timeout would come in handy?

I have started an implementation here that will generally add timeouts to streams:
https://github.com/duplicati/streamutil/blob/main/StreamUtil/TimeoutObservingStream.cs

I have prepared for this with PR #5574 which makes all backend calls `async`, except `LIST` which is what is hanging in your case. I guess this issue means we should also prioritize getting the `LIST` call over to use `IAsyncEnumerable`.

> When I add native code it shows it hanging on Socket.Receive
The stream here is supposed to be terminated by the OS when no data is flowing, but perhaps there is some keep-alive messages being sent.

> but some protection against misbehaving backends in general might also be worthwhile.

My future solution for this is to run the backend code in a separate process, so it is possible to use the operating system to stop the process.

**Comment by duplicatibot** [comment]
This issue has been mentioned on **Duplicati**. There might be relevant details there:

https://forum.duplicati.com/t/scheduled-backup-hangs-on-put-to-s3-disk-i-o-error/19164/5

**Comment by woutertinusf19** [comment]
Happy to report that using the FluentFTP backend and `--disable-piped-streaming`, backups have been stable for a month now.

**Comment by marceloduplicati** [comment]
Thank you for the feedback.
